library/common/extensions/filters/http/network_configuration/filter.cc

```diff
@@ -30,34 +30,101 @@ void NetworkConfigurationFilter::setDecoderFilterCallbacks(
+    setInfo(decoder_callbacks_->streamInfo().getRequestHeaders()->getHostValue(),
+            host_info->address());
+    return true;
+  }
+  decoder_callbacks_->sendLocalReply(Http::Code::BadRequest,
+                                     "Proxy configured but DNS resolution failed", nullptr,
+                                     absl::nullopt, "no_dns_address_for_proxy");
+  return false;
+}
+
 Http::FilterHeadersStatus
 NetworkConfigurationFilter::decodeHeaders(Http::RequestHeaderMap& request_headers, bool) {
-  const auto proxy_settings = connectivity_manager_->getProxySettings();
+  ENVOY_LOG(debug, "NetworkConfigurationFilter::decodeHeaders", request_headers);
+
+  const auto authority = request_headers.getHostValue();
+  if (authority.empty()) {
+    return Http::FilterHeadersStatus::Continue;
+  }
 
-  ENVOY_LOG(trace, "NetworkConfigurationFilter::decodeHeaders", request_headers);
+  // If there is no proxy configured, continue.
+  const auto proxy_settings = connectivity_manager_->getProxySettings();
   if (proxy_settings == nullptr) {
     return Http::FilterHeadersStatus::Continue;
   }
 
+  ENVOY_LOG(trace, "netconf_filter_processing_proxy_for_request", proxy_settings->asString());
+  // If there is a proxy with a raw address, set the information, and continue.
   const auto proxy_address = proxy_settings->address();
-
   if (proxy_address != nullptr) {
     const auto authorityHeader = request_headers.get(AuthorityHeaderName);
-    if (authorityHeader.empty()) {
-      return Http::FilterHeadersStatus::Continue;
-    }
 
-    const auto authority = authorityHeader[0]->value().getStringView();
+    setInfo(request_headers.getHostValue(), proxy_address);
+    return Http::FilterHeadersStatus::Continue;
+  }
 
-    ENVOY_LOG(trace, "netconf_filter_set_proxy_for_request", proxy_settings->asString());
-    decoder_callbacks_->streamInfo().filterState()->setData(
-        Network::Http11ProxyInfoFilterState::key(),
-        std::make_unique<Network::Http11ProxyInfoFilterState>(authority, proxy_address),
-        StreamInfo::FilterState::StateType::ReadOnly,
-        StreamInfo::FilterState::LifeSpan::FilterChain);
+  // If there's no address or hostname, continue.
+  if (proxy_settings->hostname().empty()) {
+    return Http::FilterHeadersStatus::Continue;
   }
 
-  return Http::FilterHeadersStatus::Continue;
+  // If there's a proxy hostname but no way to do a DNS lookup, fail the request.
+  if (!connectivity_manager_->dnsCache()) {
+    decoder_callbacks_->sendLocalReply(Http::Code::BadRequest,
+                                       "Proxy configured but no DNS cache available", nullptr,
+                                       absl::nullopt, "no_dns_cache_for_proxy");
+    return Http::FilterHeadersStatus::StopIteration;
+  }
+
+  // Attempt to load the proxy's hostname from the DNS cache.
+  auto result =
+      connectivity_manager_->dnsCache()->loadDnsCacheEntry(proxy_settings->hostname(), 80, *this);
```

Comment:
So we have a hostname that's a composition of a `hostname + port` and we need to specify port 80 too?

Reply:
Just checking - is it fine to keep the port hardcoded to 80 in here?